src/ForceTorqueMonitor.c

```diff
@@ -0,0 +1,180 @@
+    rosidl_runtime_c__String__Sequence__init(&g_messages_ForceTorqueMonitor.jointExternalTorque->name, g_Ros_Controller.totalAxesCount); // Number of joints in message
+
+    for (int jointIndex = 0; jointIndex < g_Ros_Controller.totalAxesCount; jointIndex += 1)
+    {
+        rosidl_runtime_c__String__assign(&g_messages_ForceTorqueMonitor.jointExternalTorque->name.data[jointIndex], g_nodeConfigSettings.joint_names[jointIndex]);
+    }
+    rosidl_runtime_c__float64__Sequence__init(&g_messages_ForceTorqueMonitor.jointExternalTorque->position, g_Ros_Controller.totalAxesCount);
+    rosidl_runtime_c__float64__Sequence__init(&g_messages_ForceTorqueMonitor.jointExternalTorque->velocity, g_Ros_Controller.totalAxesCount);
+    rosidl_runtime_c__float64__Sequence__init(&g_messages_ForceTorqueMonitor.jointExternalTorque->effort, g_Ros_Controller.totalAxesCount);
+}
+
+void Ros_ForceTorqueMonitor_Cleanup()
+{
+    rcl_ret_t ret;
+
+    Ros_Debug_BroadcastMsg("Cleanup TCP force torque publisher");
+    ret = rcl_publisher_fini(&g_publishers_ForceTorqueMonitor.forceTorqueTCP, &g_microRosNodeInfo.node);
+    if (ret != RCL_RET_OK)
+        Ros_Debug_BroadcastMsg("Failed cleaning up TCP force torque publisher: %d", ret);
+    geometry_msgs__msg__WrenchStamped__destroy(g_messages_ForceTorqueMonitor.forceTorqueTCP);
+
+    Ros_Debug_BroadcastMsg("Cleanup joint external torque publisher");
+    ret = rcl_publisher_fini(&g_publishers_ForceTorqueMonitor.jointExternalTorque, &g_microRosNodeInfo.node);
+    if (ret != RCL_RET_OK)
+        Ros_Debug_BroadcastMsg("Failed cleaning up joint external torque publisher: %d", ret);
+    sensor_msgs__msg__JointState__destroy(g_messages_ForceTorqueMonitor.jointExternalTorque);
+
+
+}
+
+void Ros_ForceTorqueMonitor_UpdateLocation()
+{
+    LONG status;
+    rcl_ret_t ret;
+
+    //Timestamp
+    INT64 theTime = rmw_uros_epoch_nanos();
+
+    Ros_Nanos_To_Time_Msg(theTime, &g_messages_ForceTorqueMonitor.forceTorqueTCP->header.stamp);
+    Ros_Nanos_To_Time_Msg(theTime, &g_messages_ForceTorqueMonitor.jointExternalTorque->header.stamp);
+
+    MP_IO_INFO registerInfo[6];
+    USHORT registerValues[6];
+    double forceTCPValues[6];
+    double torqueJointValues[6];
+
+    //Read TCP force torque data from registers
+
+    //TODO: read register addresses from params
+    registerInfo[0].ulAddr = 1000320;
+    registerInfo[1].ulAddr = 1000321;
+    registerInfo[2].ulAddr = 1000322;
+    registerInfo[3].ulAddr = 1000323;
+    registerInfo[4].ulAddr = 1000324;
+    registerInfo[5].ulAddr = 1000325;
+
+    status = mpReadIO(registerInfo, registerValues, 6);
+
+    if (status != OK)
+    {
+        Ros_Debug_BroadcastMsg("Failed to get TCP force torque data: %u", status);
+    }
+
+    //convert result to double array
+    for (int i = 0; i < 6; i += 1)
+    {
+        forceTCPValues[i] = registerValues[i];
+        forceTCPValues[i] = (forceTCPValues[i] - 10000.0) * 0.1;
+    }
+
+    //TODO: create better value assigment
```

Comment:
Would you have any ideas already on what you would consider 'better'?

Reply:
Unfortunately, no. It is possible to just leave it like that.

Reply:
What made you add that comment there? If the size and alignment of the two arrays is exactly the same, and both are contiguous, a `memcpy(..)` might work. We do that elsewhere in MotoROS2.

Reply:
Alternatively: assign the result of the conversion to the message fields directly.

Would avoid needing the temporary `forceTCPValues` array I believe.

Reply:
> What made you add that comment there? If the size and alignment of the two arrays is exactly the same, and both are contiguous, a `memcpy(..)` might work. We do that elsewhere in MotoROS2.

Yes, I wanted to use `memcpy ` there, but I do not know if the data will be converted correctly to the `geometry_msgs__msg__Wrench` format. The size and alignment of both arrays should be the same. I can test it and see if it works

Reply:
> I do not know if the data will be converted correctly to the `geometry_msgs__msg__Wrench` format.

it should just be a `double[]` in the end. You can check the `struct` definition in the generated `.h`.

Reply:
Actually, that's not true. Apologies.

It's as follows:

```c
typedef struct geometry_msgs__msg__WrenchStamped
{
  std_msgs__msg__Header header;
  geometry_msgs__msg__Wrench wrench;
} geometry_msgs__msg__WrenchStamped;
```

and then:

```c
typedef struct geometry_msgs__msg__Wrench
{
  geometry_msgs__msg__Vector3 force;
  geometry_msgs__msg__Vector3 torque;
} geometry_msgs__msg__Wrench;
```

then:

```c
typedef struct geometry_msgs__msg__Vector3
{
  double x;
  double y;
  double z;
} geometry_msgs__msg__Vector3;
```

so ideally, it will be contiguous in memory as 6 `double`s.

But I'm not sure that's guaranteed by the C99 standard.

Reply:
ok. In this case, I will just assign values directly like before but without `forceTCPValues`